pace-util/pace/util/checkpointer/snapshots.py

```diff
@@ -0,0 +1,72 @@
+import collections
+
+import numpy as np
+
+from pace.util._optional_imports import xarray as xr
+
+from .base import Checkpointer
+
+
+def make_dims(savepoint_dim, label, data_list):
+    """
+    Helper which defines dimension names for an xarray variable.
+
+    Used to ensure no dimensions have the same name but different sizes
+    when defining xarray datasets.
+    """
+    data = np.concatenate([array[None, :] for array in data_list], axis=0)
+    dims = [savepoint_dim] + [f"{label}_dim{i}" for i in range(len(data.shape[1:]))]
+    return dims, data
+
+
+class Snapshots:
+    def __init__(self):
+        self._savepoints = collections.defaultdict(list)
+        self._python_data = collections.defaultdict(list)
+
+    def compare(self, savepoint_name: str, variable_name: str, python_data):
+        self._savepoints[variable_name].append(savepoint_name)
+        self._python_data[variable_name].append(python_data)
+
+    @property
+    def dataset(self) -> "xr.Dataset":
+        data_vars = {}
+        for variable_name, savepoint_list in self._savepoints.items():
+            savepoint_dim = f"sp_{variable_name}"
+            data_vars[f"{variable_name}_savepoints"] = ([savepoint_dim], savepoint_list)
+            data_vars[f"{variable_name}"] = make_dims(
+                savepoint_dim, variable_name, self._python_data[variable_name]
+            )
+        if xr is None:
+            raise ModuleNotFoundError(
+                "xarray must be installed to use Snapshots.dataset"
+            )
+        else:
+            return xr.Dataset(data_vars=data_vars)
+
+
+class SnapshotCheckpointer(Checkpointer):
+    """
+    Checkpointer which can be used to save datasets showing the evolution
+    of variables between checkpointer calls.
+    """
+
+    def __init__(self, rank: int):
+        if xr is None:
+            raise ModuleNotFoundError(
+                "xarray must be installed to use SnapshotCheckpointer"
+            )
+        self._rank = rank
+        self._snapshots = Snapshots()
+
+    def __call__(self, savepoint_name, **kwargs):
```

Comment:
Mildly curious why this is a __call__

Reply:
I'd like to keep this as-is for this PR since it's a larger change (changing the base class as well), I'm also not sure whether it should be changed or not.

Reply:
I don't think this needs to be changed, just good to know why.